### PR TITLE
New contribs for DG

### PR DESCRIPTION
**Pull request by nmaeder**
#### What does this implement/fix? Explain your changes.

New type contribs added for Angle, Distance, Inversion and the `CrystalFF` torsion constraints. 
These are added both to the `DistGeomUtils` as well as to the `UFF/AtomTyper` (inversions)
The Corresponding tests in `ForceField` are adapted to use the new Contribs.
A `catch_tests.cpp` is added for the new contribs.

#### Any other comments?

I did not change any of the wrappers, i think this is something we need to discuss on how to achieve optimal performant code.

### CODE REVIEW

**Review by greglandrum — CHANGES_REQUESTED** [review]
Some relatively minor changes requested

File: Code/DistGeom/DistGeomUtils.cpp [hunk 1]
```diff
@@ -293,14 +295,16 @@ void addImproperTorsionTerms(ForceFields::ForceField *ff,
...
+      inversionContribs->addContrib(
+          improperAtom[n[0]], improperAtom[n[1]], improperAtom[n[2]],
           improperAtom[n[3]], improperAtom[4],
           static_cast<bool>(improperAtom[5]), forceScalingFactor);
-      ff->contribs().emplace_back(contrib);
       isImproperConstrained[improperAtom[n[1]]] = 1;
     }
   }
+  if (!inversionContribs->empty()) {
+    ff->contribs().emplace_back(inversionContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:306]: It's better to use move here:
```suggestion
    ff->contribs().emplace_back(std::move(inversionContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 2]
```diff
@@ -329,10 +335,12 @@ void addExperimentalTorsionTerms(
...
-    auto *contrib = new ForceFields::CrystalFF::TorsionAngleContribM6(
-        ff, i, j, k, l, etkdgDetails.expTorsionAngles[t].second,
-        etkdgDetails.expTorsionAngles[t].first);
-    ff->contribs().emplace_back(contrib);
+    torsionContribs->addContrib(i, j, k, l,
+                                etkdgDetails.expTorsionAngles[t].second,
+                                etkdgDetails.expTorsionAngles[t].first);
+  }
+  if (!torsionContribs->empty()) {
+    ff->contribs().emplace_back(torsionContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:343]:
```suggestion
    ff->contribs().emplace_back(std::move(torsionContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 3]
```diff
@@ -364,9 +374,11 @@ void add12Terms(ForceFields::ForceField *ff,
...
     }
     double d = ((*positions[i]) - (*positions[j])).length();
-    auto *contrib = new ForceFields::UFF::DistanceConstraintContrib(
-        ff, i, j, d - KNOWN_DIST_TOL, d + KNOWN_DIST_TOL, forceConstant);
-    ff->contribs().emplace_back(contrib);
+    distContribs->addContrib(i, j, d - KNOWN_DIST_TOL, d + KNOWN_DIST_TOL,
+                             forceConstant);
+  }
+  if (!distContribs->empty()) {
+    ff->contribs().emplace_back(distContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:381]:
```suggestion
    ff->contribs().emplace_back(std::move(distContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 4]
```diff
@@ -405,16 +421,19 @@ void add13Terms(ForceFields::ForceField *ff,
...
       double d = ((*positions[i]) - (*positions[k])).length();
-      auto *contrib = new ForceFields::UFF::DistanceConstraintContrib(
-          ff, i, k, d - KNOWN_DIST_TOL, d + KNOWN_DIST_TOL, forceConstant);
-      ff->contribs().emplace_back(contrib);
+      distContribs->addContrib(i, k, d - KNOWN_DIST_TOL, d + KNOWN_DIST_TOL,
+                               forceConstant);
     }
   }
+  if (!angleContribs->empty()) {
+    ff->contribs().emplace_back(angleContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:432]:
```suggestion
    ff->contribs().emplace_back(std::move(angleContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 5]
```diff
@@ -405,16 +421,19 @@ void add13Terms(ForceFields::ForceField *ff,
...
+  }
+  if (!distContribs->empty()) {
+    ff->contribs().emplace_back(distContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:435]:
```suggestion
    ff->contribs().emplace_back(std::move(distContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 6]
```diff
@@ -457,12 +478,13 @@ void addLongRangeDistanceConstraints(
...
         }
-        auto *contrib = new ForceFields::UFF::DistanceConstraintContrib(
-            ff, i, j, l, u, fdist);
-        ff->contribs().emplace_back(contrib);
+        distContribs->addContrib(i, j, l, u, fdist);
       }
     }
   }
+  if (!distContribs->empty()) {
+    ff->contribs().emplace_back(distContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:486]:
```suggestion
    ff->contribs().emplace_back(std::move(distContribs));
```

File: Code/DistGeom/DistGeomUtils.cpp [hunk 7]
```diff
@@ -565,14 +587,17 @@ ForceFields::ForceField *construct3DImproperForceField(
...
-      auto *contrib = new ForceFields::UFF::AngleConstraintContrib(
-          field, angle[0], angle[1], angle[2], 179.0, 180.0,
-          oobForceScalingFactor);
-      field->contribs().emplace_back(contrib);
+      angleContribs->addContrib(angle[0], angle[1], angle[2], 179.0, 180.0,
+                                oobForceScalingFactor);
     }
   }
+  if (!angleContribs->empty()) {
+    field->contribs().emplace_back(angleContribs.release());
```
Comment (greglandrum) [inline Code/DistGeom/DistGeomUtils.cpp:599]:
```suggestion
    field->contribs().emplace_back(std::move(angleContribs));
```

File: Code/ForceField/catch_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  REQUIRE(mol);
+  SECTION("absolute distance minimization") {
+    auto forceField = ForceFieldsHelper::createEmptyForceFieldForMol(*mol);
+    REQUIRE(forceField);
+    forceField->initialize();
+    auto contribs = std::make_unique<ForceFields::DistanceConstraintContribs>(
+        forceField.get());
+    contribs->addContrib(0, 1, 3.0, 3.0, 1);
+    contribs->addContrib(0, 2, 4.0, 4.0, 1);
+    forceField->contribs().emplace_back(contribs.release());
```
Comment (greglandrum) [inline Code/ForceField/catch_tests.cpp:59]:
```suggestion
    forceField->contribs().emplace_back(std::move(contribs));
```

File: Code/ForceField/catch_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  }
+  SECTION("relative distance minimization") {
+    auto forceField = ForceFieldsHelper::createEmptyForceFieldForMol(*mol);
+    REQUIRE(forceField);
+    forceField->initialize();
+    auto contribs = std::make_unique<ForceFields::DistanceConstraintContribs>(
+        forceField.get());
+    contribs->addContrib(0, 1, true, 1.0, 1.0, 1);
+    contribs->addContrib(0, 2, false, 4.0, 4.0, 1);
+    forceField->contribs().emplace_back(contribs.release());
```
Comment (greglandrum) [inline Code/ForceField/catch_tests.cpp:72]:
```suggestion
    forceField->contribs().emplace_back(std::move(contribs));
```

File: Code/ForceField/catch_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  REQUIRE(mol);
+  SECTION("absolute angle minimization") {
+    auto forceField = ForceFieldsHelper::createEmptyForceFieldForMol(*mol);
+    REQUIRE(forceField);
+    forceField->initialize();
+    auto contribs = std::make_unique<ForceFields::AngleConstraintContribs>(
+        forceField.get());
+    contribs->addContrib(0, 1, 2, 120.0, 120.0, 1);
+    contribs->addContrib(1, 2, 3, 160.0, 160.0, 1);
+    forceField->contribs().emplace_back(contribs.release());
```
Comment (greglandrum) [inline Code/ForceField/catch_tests.cpp:94]:
```suggestion
    forceField->contribs().emplace_back(std::move(contribs));
```

File: Code/ForceField/catch_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  }
+  SECTION("relative angle minimization") {
+    auto forceField = ForceFieldsHelper::createEmptyForceFieldForMol(*mol);
+    REQUIRE(forceField);
+    forceField->initialize();
+    auto contribs = std::make_unique<ForceFields::AngleConstraintContribs>(
+        forceField.get());
+    contribs->addContrib(0, 1, 2, true, 5.0, 5.0, 1);
+    contribs->addContrib(1, 2, 3, false, 160.0, 160.0, 1);
+    forceField->contribs().emplace_back(contribs.release());
```
Comment (greglandrum) [inline Code/ForceField/catch_tests.cpp:107]:
```suggestion
    forceField->contribs().emplace_back(std::move(contribs));
```

**Review by greglandrum — CHANGES_REQUESTED** [review]
The requested changes here are to use the pattern 
```
std::emplace_back(std::move(uptr));
```
instead of:
```
std::emplace_back(uptr.release());
```
I'm not sure if there's an accepted best practice, but the move, which doesn't require an extra construction of a new unique_ptr, seems like the most efficient approach.

I did a bit of benchmarking just to check and found that using move is a bit faster with either clang++ or g++ on linux while .release is a bit faster with clang++ on the mac.
So there's no obvious win.

**Comment by nmaeder** [comment]
i switched everything from 
`forcefield->contribs().emplace_back(contribs.release())`
to:
`forcefield->contribs().push_back(std::move(contribs))`

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**